### DISCUSSION TITLE
Question regarding GC notification!

### DISCUSSION THREAD

**Discussion by CodingMadness** [dotnet/dotNext#132] [Q&A] [answered]
Hey dude,

B4 I do begin asking, i am extremely astonished in what you have created with this library!! Very big congrats to you and wish ya all the best and luck with this project and life ! 

I had to say that b4 cause its litterally (next) in terms of what .net can offer . 

To my question:
Is there a way to get notified by the GC, when actual memory compression happens, so for instance when you have 
20 array elements of a Class "PlayerCharacter" and that class has 6 fields, and lets say that the GC moves at some point the array around in memory AND compresses it to reduce space, which would imply a change in <byte offset> from (itemX) to (itemX+1)
so for instance , b4 [2] were 32 bytes away from [3] and now he compacted it which makes it now only have 16 rather than 32 , as byte distance.

Is there any chance to get notified from the GC when:

1) he begins that
2) he finished that

Any idea on this much appreciated!

Cheers!

**Comment by sakno** [comment]
@Shpendicus , thanks for your feedback! GC actually compacts segments related to ephemeral generations (0 and 1). It can't compact the array. If it 20 elements long, it will stay the same during the lifetime of the array. GC can move the whole object in the memory. Additionally, there is no way to measure exact size of by-ref object (class) in the memory, because GC may add padding when object is moving from one generation to another.

P.S.: GC always apply compaction when moving objects from Gen0 to Gen1 and from Gen1 to Gen2. Gen2, by default, is not a subject of compaction, but it can be triggered explicitly from code.

**Comment by CodingMadness** [comment]
@sakno thx for ur reply! 

My main point only is actually that he does not change the distance of the fields from Item0 to ItemX, like for isntance if you do:

nint distance = ref Unsafe.ByteOffset(ref first, ref other); //----------->what will this be AFTER he has moved the entire object somewhere and ahs maybe even compressed ir or whatever it does to it.

**Comment by sakno** [comment]
There is no way to predict how the objects referenced by the array elements are placed in the memory. So the distance can be any.

**Comment by sakno** [comment]
If you really care about data locality (due to performance reasons), convert `PlayerCharacter` to value type. In that case, all array elements will occupy contiguous memory block, instead of references (pointers) to the objects.

**Comment by CodingMadness** [comment]
> If you really care about data locality (due to performance reasons), convert `PlayerCharacter` to value type. In that case, all array elements will occupy contiguous memory block, instead of references (pointers) to the objects.

Hmm, i cant do that cause i am building a datastructure as a wrapper for other datastructures/types so i kinda have to work on what I do get and i dont wanna constraint that, cause it should work (semtantically) for both. 

And it is okay when the distance is any value, the main criteria only is that the distance does not change randomly at runtime!!

This I ask you, if that can happen, or is the layout fixed once he created the entire array?

**Comment by CodingMadness** [comment]
![memory compaction](https://user-images.githubusercontent.com/28008874/202825769-ff525b73-6e8a-4af9-97c3-76582f2f12c5.png)

**Comment by CodingMadness** [comment]
> > If you really care about data locality (due to performance reasons), convert `PlayerCharacter` to value type. In that case, all array elements will occupy contiguous memory block, instead of references (pointers) to the objects.
> 
> Hmm, i cant do that cause i am building a datastructure as a wrapper for other datastructures/types so i kinda have to work on what I do get and i dont wanna constraint that, cause it should work (semtantically) for both.
> 
> And it is okay when the distance is any value, the main criteria only is that the distance does not change randomly at runtime!!
> 
> This I ask you, if that can happen, or is the layout fixed once he created the entire array?

Also it is OKAY if he compacts, but i would love to get notified for this! 

So, either:

1) dont compact at all (maybe this you can force for certain objects?)
2) If you compact, notify me, so i can recompute the offsets!

**Comment by sakno** [comment]
Ok, I got it. But GC doesn't change the layout of array elements. Address of each array element is aligned in memory according to its size. For ref types (class), each element is always aligned to `sizeof(nint)`, i.e. 4 bytes on 32-bit machine or 8 bytes on 64-bit machine. As a result, if address of `array[0]` is some `X`, then address of n-th element will be `Offset = X+sizeof(nint)*n` and `Offset mod sizeof(nint) = 0`

To detect heap compaction, you can do the following:
```csharp
await GCNotification.HeapCompaction().WaitAsync();
```

**Comment by CodingMadness** [comment, marked as answer]
> Ok, I got it. But GC doesn't change the layout of array elements. Address of each array element is aligned in memory according to its size. For ref types (class), each element is always aligned to `sizeof(nint)`, i.e. 4 bytes on 32-bit machine or 8 bytes on 64-bit machine. As a result, if address of `array[0]` is some `X`, then address of n-th element will be `X+sizeof(nint)*n`.
> 
> To detect heap compaction, you can do the following:
> 
> ```cs
> await GCNotification.HeapCompaction().WaitAsync();
> ```

ahh thx mate good to know

**Reply by sakno** [reply]
I converted issue to discussion. Please mark it as answered if we are finished.

**Reply by CodingMadness** [reply]
okay last question to this:

consider this example: 

```
Note:

the indeed more relevant info is if the GC ever compacts the room between array of classes/structs like:

class Test {
  public int A;
  public double B;
  public string C;
}

Test[] tests = new[20]; 

ref char first = ref Unsafe.AsRef(tests[0].C[0]);
ref char next = ref Unsafe.AsRef(tests[1].C[0]);

nint diff = Unsafe.ByteOffset(ref first, ref next); //can this value be different when GC compacts or does he eveb compact such?
```

can i compute that only 1 time and it will be fixed?? or can that "diff" change by the GC?

**Reply by sakno** [reply]
No. You trying to measure memory location of two strings that are ref types (classes). Two strings are two different objects, they can be placed by GC at any memory location in the heap. Moreover, in some point in time, one of the object can be in Gen0, another one in Gen1. I'm just curious, what you trying to achieve?

**Reply by sakno** [reply]
Even GC notification will not help you. Because you receive notification about GC happened at some point in time in the past. It guarantees `happens-before` semantics, but you don't know when it was happened actually.

**Reply by CodingMadness** [reply]
> No. You trying to measure memory location of two strings that are ref types (classes). Two strings are two different objects, they can be placed by GC at any memory location in the heap. Moreover, in some point in time, one of the object can be in Gen0, another one in Gen1. I'm just curious, what you trying to achieve?

I am trying to know where begins the "char" of the string of the array, and where is the next char of the next element in the array, i need to know these because i m writing a specialized buffer for our company where you need to create chunks based on array inputs and these arrays are array of classes and these classes do have strings as fields.

so i need to know each beginning of each string in the array. which i am computing successfully, i assumed arrays are contigoeus regions of size in heap and the GC moves and expands the entire thing and not randomly some of it?  

Cant I just pin the entire array for the duration of the internal chunk creator i am doing?

**Reply by sakno** [reply]
Array allocation is not correlated with allocation of strings in your case. String itself is not in the array, the array contains only references to that strings (pointers). So there is no correlation between address of an array and the strings. Pinning of the array doesn't pin the underlying elements in memory. Pinned instances of `PlayerCharacter` do not pin the strings they reference. All those things are independent objects, coupled through strong references, nothing more.

**Reply by CodingMadness** [reply]
what can i do then to get what i want? just fixing the string ref then? with a GC handle?

**Reply by sakno** [reply]
With that, you will have a high fragmentation of managed memory heap. Pinned objects cannot be moved by GC. Even if you have two pinned strings placed in contiguous memory block, you can't read the contents of two strings using simple pointer arithmetic, because those strings are two different objects. .NET has [Object Header](https://devblogs.microsoft.com/premier-developer/managed-object-internals-part-2-object-header-layout-and-the-cost-of-locking/), a few bytes at the beginning of each object with internal information about its type, synchronized block, etc. In other words, you will get a some portion of raw bytes unrelated to string contents.

Try to re-think the architecture for your buffer. For instance, place all strings in some rented buffer which is available in .NET and .NEXT both.

**Reply by CodingMadness** [reply]
Hmm, but does it not be costly to copy the strings into that buffer? i mean i do not have control of where the array comes from, i just know the content but not its src and such so i do not know if its a good idea to copy the entir array into a buffer?

**Reply by sakno** [reply]
We returning back to the question: what you trying to achieve? Do you want to copy N strings located randomly in memory as a contiguous buffer? If yes, you need to copy those strings in contiguous memory. Otherwise, you cannot achieve this. But copy overhead probably will be the same (or even larger) than just iterating over strings (and copy each string as a block of memory). The better way to check this is to provide benchmarks.

**Reply by CodingMadness** [reply]
I kinda followed your buffer advise, thank you!

**Reply by sakno** [reply]
I recommend to take a look at [this](https://dotnet.github.io/dotNext/features/io/buffers.html) page with information about buffers.
